### PR TITLE
feat(pie-checkbox): DSW-1574 checkbox functionality

### PR DESCRIPTION
**Pull request by dandel10n**
## Describe your changes (can list changeset entries if preferable)

Checkbox component main functionality.
Storybook feature url: https://pr1437-storybook.pie.design/?path=/story/checkbox--default

**Note:** styles, accessibility requirements and form integration are out of scope for this PR.

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] I have added thorough tests where applicable (unit / component / visual)
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview

### Reviewer 1 – @ashleynolan 
- [x] I have reviewed the `PIE Storybook` PR preview

### Reviewer 2 - Fernando
- [x] I have reviewed the `PIE Storybook` PR preview

### CODE REVIEW

File: apps/pie-storybook/stories/pie-checkbox.stories.ts [hunk 1]
```diff
@@ -1,36 +1,109 @@
...
+            },
+        },
+        disabled: {
+            description: 'If true, disables the checkbox.',
+            control: 'boolean',
+            defaultValue: {
+                summary: false,
+            },
+        },
+        indeterminate: {
```
Comment (ashleynolan) [inline apps/pie-storybook/stories/pie-checkbox.stories.ts:62]: This is super interesting – I had seen this state used but didn't know the full use case of it until now ([and read this article too](https://css-tricks.com/indeterminate-checkboxes/)).

File: apps/pie-storybook/stories/pie-checkbox.stories.ts [hunk 2]
```diff
@@ -1,36 +1,109 @@
...
+            description: 'Indicates whether the checkbox visually shows a horizontal line in the box instead of a check/tick. It has no impact on whether the checkbox\'s value is used in a form submission. That is decided by the checked state, regardless of the indeterminate state.',
+            control: 'boolean',
+            defaultValue: {
+                summary: false,
+            },
+        },
+    },
     args: defaultArgs,
     parameters: {
+        layout: 'centered',
```
Comment (ashleynolan) [inline apps/pie-storybook/stories/pie-checkbox.stories.ts:72]: Out of interest – how does this visually change the preview (compared with the default?)

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 3]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+        test.describe('value', () => {
+            test('should default to "on" string if no value prop provided', async ({ mount }) => {
+                // Arrange
+                const component = await mount(PieCheckbox, {});
+
+                // Act
+                const checkbox = component.locator(componentSelector);
+
+                // Assert
+                expect((await checkbox.inputValue())).toBe('on');
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:38]: Potentially stupid question – why would the default value be `on` and not `undefined` or `''` (empty) here?

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 4]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+                    const component = await mount(PieCheckbox, {});
+
+                    // Act
+                    const checkbox = component.locator(componentSelector);
+
+                    // Assert
+                    expect(checkbox).not.toBeDisabled();
+                });
+
+                test('should still be able to focus the component', async ({ page }) => {
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:127]:
```suggestion
                test('should be able to focus the component', async ({ page }) => {
```

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 5]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+                    await checkbox.focus();
+
+                    // Assert
+                    expect(checkbox).toBeFocused();
+                });
+            });
+        });
+
+        test.describe('required', () => {
+            test('should not render a required attribute if no required provided', async ({ mount }) => {
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:142]:
```suggestion
            test('should not render a required attribute if the required prop isn’t provided', async ({ mount }) => {
```

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 6]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+                });
+
+                // Act
+                const isValid = await page.evaluate(() => document.querySelector('pie-checkbox')?.validity.valid);
+
+                // Assert
+                expect(isValid).toBe(false);
+            });
+
+            test('should be valid state if checkbox is checked and required', async ({ mount, page }) => {
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:183]:
```suggestion
            test('should be in a valid state if checkbox is checked and required', async ({ mount, page }) => {
```

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 7]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+                });
+
+                // Act
+                const checkbox = component.locator(componentSelector);
+
+                // Assert
+                expect((await checkbox.getAttribute('required'))).toBe('');
+            });
+
+            test('should be in invalid state if checkbox is unchecked and required', async ({ mount, page }) => {
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:168]:
```suggestion
            test('should be in an invalid state if checkbox is unchecked and required', async ({ mount, page }) => {
```

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts [hunk 8]
```diff
@@ -17,4 +24,194 @@ test.describe('PieCheckbox - Component tests', () => {
...
+
+                // Act
+                await component.locator(componentSelector).click();
+                const isValid = await page.evaluate(() => document.querySelector('pie-checkbox')?.validity.valid);
+
+                // Assert
+                expect(isValid).toBe(true);
+            });
+
+            test('should be valid state if the input has a checked prop and required', async ({ mount, page }) => {
```
Comment (ashleynolan) [inline packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:200]:
```suggestion
            test('should be in a valid state if the input has a checked prop and required', async ({ mount, page }) => {
```

**Review by ashleynolan — COMMENTED** [review]
Looks good – few super small things, but LGTM

File: apps/pie-storybook/stories/pie-checkbox.stories.ts (hunk 1 above)
Comment (fernandofranca) [reply apps/pie-storybook/stories/pie-checkbox.stories.ts:62]: Was asking myself what `indeterminate` was for a few seconds ago! 😄

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts (hunk 3 above)
Comment (dandel10n) [reply packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts:38]: Not stupid at all, I forgot to add a note about it as I learned about it myself just while working on this component. It is a default value of `value` attr for html input type=checkbox so if it is not set by a user when the form is submitted with the checked checkbox the submitted value is "on". [The mdn doc](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#value) mentioning it for the ref.

File: apps/pie-storybook/stories/pie-checkbox.stories.ts (hunk 2 above)
Comment (dandel10n) [reply apps/pie-storybook/stories/pie-checkbox.stories.ts:72]: Actually I checked that this parameter doesn't change the component's layout. For icon stories it centers the icon vertically but checkbox stays centered even without it

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts (hunk 3 above)
Comment (ashleynolan) [reply packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts]: Might be worth adding a comment into the test linking to that, as that's really useful to know 😄

**Review by fernandofranca — APPROVED** [review]

File: packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts (hunk 3 above)
Comment (dandel10n) [reply packages/components/pie-checkbox/test/component/pie-checkbox.spec.ts]: I updated the test and added a comment to the `expected` constant.

**Pull request merged**